### PR TITLE
feat(sql): less/greater correlated subqueries support for timestamp column

### PR DESCRIPTION
**Pull request by puzpuzpuz**
Now it's possible to write queries like the following:
```sql
select *
from x
where ts > (select min(ts) from x)
  and ts < (select max(ts) from x);
```

In case if `ts` is the designated timestamp, time filter intrinsics will kick in and apply the filter to the scanned partitions. This is also applied to the `ts = (<subquery>)` operator we already have.

Also fixes missing `Function#close()` calls in case of exceptions in `WhereClauseParser`.

### CODE REVIEW

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 374 / 407 (91.89%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/eq/EqTimestampCursorFunctionFactory.java|38|44|86.36%|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|133|146|91.10%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/lt/LtTimestampCursorFunctionFactory.java|83|90|92.22%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/lt/GtTimestampCursorFunctionFactory.java|83|90|92.22%|
|:large_blue_circle:|io/questdb/griffin/FunctionFactoryCache.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InSymbolCursorFunctionFactory.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/lt/LtTimestampFunctionFactory.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/RuntimeIntervalModel.java|12|12|100.00%|

File: core/src/main/java/io/questdb/griffin/engine/functions/lt/GtTimestampCursorFunctionFactory.java [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+        }
+
+        switch (metadata.getColumnType(0)) {
+            case ColumnType.TIMESTAMP:
+            case ColumnType.NULL:
+                return new TimestampCursorFunc(factory, args.getQuick(0), args.getQuick(1));
+            case ColumnType.STRING:
+                return new StrCursorFunc(factory, args.getQuick(0), args.getQuick(1), argPositions.getQuick(1));
+            case ColumnType.VARCHAR:
+                return new VarcharCursorFunc(factory, args.getQuick(0), args.getQuick(1), argPositions.getQuick(1));
```
[the remark this answers is not in the document]
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/griffin/engine/functions/lt/GtTimestampCursorFunctionFactory.java:89]: Yes, that's correct: string and varchar functions may fail when parsing the timestamp string.

File: core/src/main/java/io/questdb/griffin/WhereClauseParser.java [hunk 2]
```diff
@@ -1678,22 +1765,33 @@ private void excludeKeyValue(IntrinsicModel model, FunctionParser functionParser
...
+            if (funcType == ColumnType.SYMBOL
+                    || funcType == ColumnType.STRING
+                    || funcType == ColumnType.CHAR
+                    || funcType == ColumnType.UNDEFINED
+                    || funcType == ColumnType.NULL
+                    || funcType == ColumnType.VARCHAR) {
+                return Chars.toString(func.getStrA(null));
+            } else {
+                throw SqlException.$(node.position, "Unexpected function type [").put(ColumnType.nameOf(funcType)).put("]");
+            }
```
[the remark this answers is not in the document]
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/griffin/WhereClauseParser.java:1792]: Do you think switch will be more readable?

[the remark this answers is not in the document]
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/griffin/WhereClauseParser.java:1792]: To me, they read equally.

**Comment by puzpuzpuz** [comment]
@nwoolmer thanks for the review!

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 373 / 406 (91.87%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/eq/EqTimestampCursorFunctionFactory.java|37|43|86.05%|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|133|146|91.10%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/lt/LtTimestampCursorFunctionFactory.java|83|90|92.22%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/lt/GtTimestampCursorFunctionFactory.java|83|90|92.22%|
|:large_blue_circle:|io/questdb/griffin/FunctionFactoryCache.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InSymbolCursorFunctionFactory.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/lt/LtTimestampFunctionFactory.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/RuntimeIntervalModel.java|12|12|100.00%|

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 373 / 406 (91.87%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/eq/EqTimestampCursorFunctionFactory.java|37|43|86.05%|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|133|146|91.10%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/lt/LtTimestampCursorFunctionFactory.java|83|90|92.22%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/lt/GtTimestampCursorFunctionFactory.java|83|90|92.22%|
|:large_blue_circle:|io/questdb/griffin/FunctionFactoryCache.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InSymbolCursorFunctionFactory.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/lt/LtTimestampFunctionFactory.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/RuntimeIntervalModel.java|12|12|100.00%|

**Pull request merged**